File: Coursera/AlgorithmOnGraphs/week2/1_acyclicity.py
"""
## Determining an Order of Courses

# Problem Introduction
    Now, when you are sure that there are no cyclic dependencies in the given CS curriculum, you would like to
    find an order of all courses that is consistent with all dependencies. For this, you find a topological ordering
    of the corresponding directed graph.

# Problem Description
    Task.           Compute a topological ordering of a given directed acyclic graph (DAG) with 𝑛 vertices and 𝑚 edges.
    Input Format.   A graph is given in the standard format.
    Constraints.    1 ≤ 𝑛 ≤ 105, 0 ≤ 𝑚 ≤ 105. The given graph is guaranteed to be acyclic.
    Output Format.  Output any topological ordering of its vertices. (Many DAGs have more than just one
                    topological ordering. You may output any of them.)

# Sample 1.
Input:
4 3
1 2
4 1
3 1
Output:
4 3 1 2

# Sample 2.
Input:
4 1
3 1
Output:
2 3 1 4

# Sample 3.
Input:
5 7
2 1
3 2
3 1
4 3
4 1
5 2
5 3
Output:
5 4 3 2 1
"""

import sys


def acyclic(adj):
    return 0

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
    print(acyclic(adj))

File: Coursera/AlgorithmOnGraphs/week2/1_Acyclicity.java
"""
## Checking Consistency of CS Curriculum

# Problem Introduction
    A Computer Science curriculum specifies the prerequisites for each course as a list of courses that should be
    taken before taking this course. You would like to perform a consistency check of the curriculum, that is,
    to check that there are no cyclic dependencies. For this, you construct the following directed graph: vertices
    correspond to courses, there is a directed edge (𝑢, 𝑣) is the course 𝑢 should be taken before the course 𝑣.
    Then, it is enough to check whether the resulting graph contains a cycle.

# Problem Description
    Task.           Check whether a given directed graph with 𝑛 vertices and 𝑚 edges contains a cycle.
    Input Format.   A graph is given in the standard format.
    Constraints.    1 ≤ 𝑛 ≤ 10^3, 0 ≤ 𝑚 ≤ 10^3.
    Output Format.  Output 1 if the graph contains a cycle and 0 otherwise.

# Sample 1
Input:
4 4
1 2
4 1
2 3
3 1

Output:
1

# Sample 2
Input:
5 7
1 2
2 3
1 3
3 4
1 4
2 5
3 5

Output:
0
"""

import java.util.ArrayList;
import java.util.Scanner;

public class Acyclicity {
    private static int acyclic(ArrayList<Integer>[] adj) {
        //write your code here
        return 0;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
        }
        System.out.println(acyclic(adj));
    }
}

